### DISCUSSION TITLE
Different shifter model need advice :)

### DISCUSSION THREAD

**Discussion by NikderNoob95** [azzajess/USB-Shifter-to-Thrustmaster-Wheelbase#1] [Q&A]
Hi, what a cool idea. I've bought an alibaba shifter 6+R similar like yours and want to connect it to a T300RS wheel to get it to work
with Gran Turismo 7 on PS5.
So I found your page and soldered everything like you described, the hardware part is working now I think.
But after studying the code I found out, that my shifter looks exactly the same, but has a switch for each gear instead of
a Joystick/axis.
Do you have an idea how I can get it to detect the button inputs to shift the gears? I have no changing signal in the serial plotter now...
Thank you

**Comment by azzajess** [comment]
Hey, Give me a bit of time to look into this.
First we would have to debug what commands that are sent from the usb shifter to the arduino. So instead of the debug for controller that we have currently, we would need something that can read keyboards commands. Then we can note that down.
Then once we have those commands of each gear/button of the shifter, then make adjust the code so it translates them to the th8a commands.
Anyway, I have an idea to implement, just got to get the time to test some things out.

**Reply by NikderNoob95** [reply]
Ahh yes, glad to hear there is a chance. Sure, take your time, I wouldn't even know where to start so... Very interesting for me. I will look constantly for news, if I can help with debugging or so I will do that.

**Reply by azzajess** [reply]
Still going through this, so im going to post a bit at time in between other work.

https://github.com/felis/USB_Host_Shield_2.0/blob/master/examples/HID/USBHIDBootKbd/USBHIDBootKbd.ino
Flash this onto the Arduino (without shifter connected)

Then once flashed and still connected to computer, open serial monitor in Arduino and plug in the shifter.
Then go down through each shift, making note of the different commands
After you go through the whole lot, copy and paste it here and ill see what I get from it. just also mention what each of the shift is and their corresponding letter. Let me know if you have any questions

eg
![image](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/40465008/18a40e0c-10e2-4e13-932a-d05151e620b6)

**Reply by NikderNoob95** [reply]
Okay I flashed the USBHIDBootKbd.ino on the Arduino and connected the shifter, but the serial plotter only shows "Start"... I tried resetting it and to change the power source directly 3.3v from the wheel, but the same thing. I also connected my keyboard but nothing changes.

**Reply by NikderNoob95** [reply]
But there is a difference, when I plug in my keyboard, the red LED flashes only short and goes out, when I connect the shifter it blinks and stays on. Maybe my keyboard needs more voltage? However there are no Inputs detected from the shifter or keyboard...

**Reply by NikderNoob95** [reply]
If I connect the shifter in windows, it is listed as an gamecontroller in the devices menu. As an Xbox Controller in the settings. And the controller settings show numbers from 1 to 12 and the gears are 1 to 8, while 7 is left out. Maybe this is somehow helpful?

**Reply by azzajess** [reply]
Oh, then if thats the case it is a game controller and not a keyboard, therefore this script wont work anyway, because its waiting for keyboard signals not game controller signals. I thought when you plugges in the shifter, it would send keyboard commands when you change the shift.

Just to clarify what I meant, you were to have the shifter connected to the programmer while the programmer is connected between the pc and pro mini board. So with that said, I wanted to ask if you tried the original script, but uncommented the section where it mentions for debugging.
So uncomment the below inside the main USBShifter2PS2.ino script, flash it, and while the programmer is still connected to the pc, open serial monitor and plug in the shifter. Then move the shifter around and note the values changed.

![image](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/40465008/1aa636d8-67d9-4e5a-85db-8e5ccf332713)

Or am I misunderstanding things :o

**Reply by NikderNoob95** [reply]
Ohh i will try this now. I had a USB Dongle with 5v. The red flashing LED is new since I tried the wheel as power source. Maybe it's my fault, I will update you soon.

**Reply by NikderNoob95** [reply]
I tried it now but with the original file with uncommented debugging it isnt working also so false alarm I think... If I look inside the shifter there are metal plates at each gear point so I assumed it has switches instead of a joystick-like controller. Maybe its somehow an other Xbox Controller? I don't know what they could have built in there...

The serial monitor shows:

Start
Gear 0

But nothing happens when I shift the gears.

**Reply by azzajess** [reply]
The metal plates are micro switches which are directly connected to some micro controller that acts as a controller over USB. Although now im wondering if its being detected as an xinput device or something else....

This is the 3v3 Arduino currently connected to the shifter yeah and you have also disconnected and reconnected the shifter while its connected to serial monitor etc?

Try flashing these files and let me know if you get any data when moving the shifter. 
https://github.com/felis/USB_Host_Shield_2.0/tree/master/examples/HID/USBHIDJoystick

I will also get additional info from my shifter to compare with yours and see if there is a point of difference, but wont be for a bit.

**Reply by NikderNoob95** [reply]
Yes I have the 3.3V Arduino with the Atmega328P CPU and tried disconnecting and reconnecting the shifter while in serial monitor. I will try the files later. My shifter looks nearly exactly like in your picture, but there is no wire coming out of it what is connected to the shifting stick...

**Reply by NikderNoob95** [reply]
I now have copied all 3 files into a folder, opened them and flashed the .ino, checked the serial monitor and sadly there are no inputs, too. 

You are right, behind the switches are microcontrollers/little platines...

**Comment by azzajess** [comment]
Can you give me a screenshot of this?
![image](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/40465008/1f0a3983-4262-4913-a1ed-38956c141923)
And any screenshots of where it says xbox controller? Anything that says xinput or dinput for example?
Im thinking its not like a generic USB controller like mine and something more akin to an xinput or something

**Reply by NikderNoob95** [reply]
Somehow yesterday it was listed as "Controller" and as an Xbox Controller in the settings but I think I mistaked a virtual controller I made some day sorry man... Today I discovered it is listed as an Generic USB Joystick. My bad... :/ 
![Controller Settings USB Shifter](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/137695361/9ae64ca4-06b6-4ec9-9191-e8ef50f81097)

**Reply by NikderNoob95** [reply]
And strange that it is the same as yours... I really hope it's not my arduino what is causing the problem here. Shifter works okay in windows...

**Reply by NikderNoob95** [reply]
So it definitively was my fault with the xbox controller, clicked on it by mistake and it has the same sheme with the 12 buttons. Forget this...

**Reply by azzajess** [reply]
Its possible? You mentioned something about a 5v dongle? what did you mean by this?

The Arduino is posting code with the gear 0 display stuff when plugged in, so its probably working but, what im not sure about, is the usb shield?
Run the below script and plug in a run of the mill membrane usb keyboard, something that doesnt require a lot of current. Then just press some random buttons on the keyboard and see if you actually get any feedback in the serial monitor. Let me know how you go. Maybe the usb shield is properly connected to the Arduino?
https://github.com/felis/USB_Host_Shield_2.0/blob/master/examples/HID/USBHIDMultimediaKbd/USBHIDMultimediaKbd.ino

**Reply by NikderNoob95** [reply]
Yes I didn't know that you need a 3.3v dongle because I've read the arduino has a current converter. Maybe that info was wrong but I'll try the Kbd file now...

**Reply by NikderNoob95** [reply]
Aah there are no inputs. Tried it with another keyboard, Led on Arduino stays on but on the Keyboard nothing happens stays dark... So thats the problem. Do you know what I can do now?

**Reply by azzajess** [reply]
From memory the 3v3 requirement is based off the fact the wheel outputs 3v3 and the USB shield data logic requires 3v3 volts.

> Mini shield The Mini is 3.3V-only device; it is designed to mate with Arduino Pro Mini.
&
> 3.3V. Power pin of the shield. Do not supply 5V to this pin – MAX3421E won’t be able to detect devices on VBUS and may also be damaged!

This is from https://chome.nerpa.tech/usb-host-shield-hardware-manual/ if you want to read

**Reply by NikderNoob95** [reply]
Okay sh*t :( Then maybe I have fried it. But really thank you very much for your help. I maybe have to buy another one... If you know something how I could save this let me know.

**Reply by azzajess** [reply]
So just to clarify, you have a pro mini 3v3 connected to the pc with a 3v3 programmer? supplying 3v3 to the Arduino?
Try un plugging the keyboard and plugging it back to see if commands suddenly work. I had a similar issue where the lights on keyboard wont work but i still got commands.

**Reply by NikderNoob95** [reply]
I have the CP2102 USB to TTL converter but only connected the data lines Txd  and Rxd, power I take now over the Din Port of the wheel, could this be an issue too? Then I will order a new programmer with the right voltage...

**Reply by NikderNoob95** [reply]
F I now found my converter has a 3v3 port. Wow I will try this.

**Reply by NikderNoob95** [reply]
Okay that doesn't work either, I'm very new to this and first didnt think of 3v3 being a 3.3v port and didn't look at this pin on the adapter again. So dumb but learned this the hard way now... 

Yes I unplugged and plugged in the keyboard in the serial monitor such as the shifter with the original file flashed. Maybe my USB Host Shield is defect now.

**Reply by azzajess** [reply]
Thats possible, pushing 5v to the chip. So that means the Arduino pro mini you have is 3v3 8mhz or 5v 16mhz? 
Also I wouldn't recommend having it connected between the wheel and pc, so when testing it should all be run from the pc.
pc usb to ttl with 3v3 out to Arduino 3v3,gnd,tx,rx and is also attached to the shield.

You can also try the script with the uncommented debug stuff to see if the codes are different to mine? hoping the board isnt stuffed, now that we are running 3v3 into the Arduino.

**Reply by NikderNoob95** [reply]
Yes I have the 3v3 8mhz arduino. I tried your code with the uncommented debug stuff but I still get no inputs with the shifter connected to the host shield and the correct wiring of the Usb to TTL converter. I think I will order a new Hosts Shield and try this before ordering a new Arduino. Then I will report if it works.

**Reply by NikderNoob95** [reply]
The Board did work with 5v USB power so I didn't notice there was something wrong with my wiring, arduino took the uploads and was responding so I hope at least the arduino survived. The USB Host Shield did nothing with 5v connection and seemed to work when I gave the Board 3.3v. But never saw any Signal from it besides the red flashing LED when I connect something...

**Reply by azzajess** [reply]
That sucks :( 
I hope the new board will solve the problem. 
Maybe there is some code available that will help further debug if there is an issue with the shield while you wait for a new one? I don't actually know if something like that exists :(
Sorry to hear you had troubles though

**Reply by NikderNoob95** [reply]
Yes I will look if I can do something to fix this. Maybe someone had the same problem. If I got any news I will post it here, I want it to work now. :D But really thank you for this effort it definitively helped me to understand more.

**Reply by azzajess** [reply]
I did come across this which may help?
https://chome.nerpa.tech/mcu/arduino/usb-shield/troubleshooting-arduino-usb-host-shield/

**Comment by NikderNoob95** [comment]
Hi, I got my new USB Host Shield today and yes, I now get the inputs from my USB Keyboard so partially there is a success yay! :D Bot when I flash the Shifter file with the uncommented debug I still have nothing. So maybe the idea that my shifter is somehow different wasn't wrong... What can we do?

**Reply by NikderNoob95** [reply]
Tried all the files but no result, could it be that my shifter needs 5v to work? My "big" keyboard doesn't work, too. No flashing red LED on the arduino, the other keyboard works perfectly and LED flashes and stays on, so with the shifter, but that thing will not get any inputs...

**Reply by NikderNoob95** [reply]
But I think the shifter is detected by the arduino if I plug it in. Led does the same as with the functioning keyboard.

**Reply by azzajess** [reply]
I'll look into this more tomorrow, but for the moment, I don't think your shifter is 5v at this stage. I'm not convinced it would any different really since mentioning it's like a game controller in windows.

The LEDs flashing on audrino are not the indicator to know your getting data, it's the debug output on the serial monitor. I have to double check this myself but are you getting like a debug start message (or any message) in serial after you reset audrino whilst it's connected to PC?

Take a few pictures of what your audrino + USB shield and how you have it connected.

Also I have made some updates to the main page of the GitHub that elaboratesthr procedure if you haven't already come across it.

**Reply by NikderNoob95** [reply]
Okay, yes if I restart the arduino the serial plotter says "start" and when I connect something it plots "start" again... This is how I connected the arduino:
![IMG_20230708_111501](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/137695361/77ac1235-1e48-4de0-b97c-b9a6759a7fdc)

**Reply by azzajess** [reply]
So after some testing, i too get the same thing when plugging in some keyboards, nothing comes up. A numpad i have works and actually shows up the information. Also tried a mouse and that too showed up. Not sure why some devices do and some dont....
Did you try this Arduino code? It should just come up with info regarding usb device connected. It didnt for one of my keyboards but did for the mouse, numpad and shifter. Just make sure that once flashed, close and re open serial monitor and press the reset button when the shifter is connected. Let me know if any info shows up in the serial plotter. Looking for any info from any usb device, just to confirm that we get something out of a usb shield.
https://github.com/felis/USB_Host_Shield_2.0/tree/master/examples/USB_desc

Have to look into why some devices work...

**Reply by NikderNoob95** [reply]
No I didn't try this. If I connect the shifter I get this information:

Start


01
--

Device descriptor: 
Descriptor Length:	12
Descriptor type:	01
USB version:		0100
Device class:		00
Device Subclass:	00
Device Protocol:	00
Max.packet size:	08
Vendor  ID:		0079
Product ID:		0006
Revision ID:		0107
Mfg.string index:	01
Prod.string index:	02
Serial number index:	00
Number of conf.:	01

Configuration descriptor:
Total length:		0029
Num.intf:		01
Conf.value:		01
Conf.string:		00
Attr.:			80
Max.pwr:		FA

Interface descriptor:
Intf.number:		00
Alt.:			00
Endpoints:		02
Intf. Class:		03
Intf. Subclass:		00
Intf. Protocol:		00
Intf.string:		00
Unknown descriptor:
Length:		09
Type:		21
Contents:	100121012265000705

Endpoint descriptor:
Endpoint address:	82
Attr.:			03
Max.pkt size:		0008
Polling interval:	0A

Endpoint descriptor:
Endpoint address:	02
Attr.:			03
Max.pkt size:		0008
Polling interval:	0A


Addr:1(0.0.1)

**Reply by azzajess** [reply]
Thats good since they are indeed communicating via usb shield
This is what got with my shifter for comparison.

```
01
--

Device descriptor:
Descriptor Length:    12
Descriptor type:    01
USB version:   	 0100
Device class:   	 00
Device Subclass:    00
Device Protocol:    00
Max.packet size:    08
Vendor  ID:   	 0079
Product ID:   	 0006
Revision ID:   	 0107
Mfg.string index:    01
Prod.string index:    02
Serial number index:    00
Number of conf.:    01

Configuration descriptor:
Total length:   	 0029
Num.intf:   	 01
Conf.value:   	 01
Conf.string:   	 00
Attr.:   		 80
Max.pwr:   	 FA

Interface descriptor:
Intf.number:   	 00
Alt.:   		 00
Endpoints:   	 02
Intf. Class:   	 03
Intf. Subclass:   	 00
Intf. Protocol:   	 00
Intf.string:   	 00
Unknown descriptor:
Length:   	 09
Type:   	 21
Contents:    100121012265000705

Endpoint descriptor:
Endpoint address:    81
Attr.:   		 03
Max.pkt size:   	 0008
Polling interval:    0A

Endpoint descriptor:
Endpoint address:    02
Attr.:   		 03
Max.pkt size:   	 0008
Polling interval:    0A


Addr:1(0.0.1)
```
Pretty much all of it is the same with the exception of endpoint address being 82 on yours, but i don't think that's the reason....

So now begs the issue of the shifter is communicating why isnt it coming up in debug. Which type of code did you debug? This one or the one liner?
```  Serial.print("X: ");
  Serial.print(evt->x);
  Serial.print(" Y: ");
  Serial.print(evt->y);
  Serial.print(" Hat Switch: ");
  Serial.print(evt->hat);
  Serial.print(" Twist: ");
  Serial.print(evt->twist);
  Serial.print(" Slider: ");
  Serial.print(evt->slider);
  Serial.print(" Buttons: ");
  Serial.print(evt->buttons);
  Serial.println();
```

Also I dont know why but sometimes after i flash something in Arduino, i get some sort of error and have to close the serial monitor and re open it, reset Arduino and then the serial monitor works...

Let me know how that goes

**Reply by NikderNoob95** [reply]
I tried uncommenting the full code, the oneliner and closing and reopening the serial monitor, reconnected the shifter several times but there are no outputs in the serial monitor than Start, Gear 0... I will try it again now but I'm stuck here, will update you if I get any input.

**Reply by azzajess** [reply]
This is really puzzling. My lack of knowledge on all this doesn't help fault find  the issue either :(
Now that you have a working ish hosts shield, did the 
https://github.com/felis/USB_Host_Shield_2.0/tree/master/examples/HID/USBHIDJoystick
Give back any results?

Also if you really want this to work, there is another method but you loose access to using the shifter on oc via USB. 
It essentially involves directly connecting the microswitches to the audrino and using that to then trigger PS2 commands. That's a little more work. I haven't tried it but in theory it would work. Requires some code rewrite.

**Reply by NikderNoob95** [reply]
Okay interesting, I will try the USBHIDJoistick file now and the other files, maybe something happens... If not I will disassemble the shifter and try this, sounds like that is a better way. Nice :D

**Reply by NikderNoob95** [reply]
Nope, nothing.

**Comment by NikderNoob95** [comment]
This is what I got out of it. I made a picture of the USB platine, too, maybe this is interesting to compare. 

![shifter (1)](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/137695361/87d90627-f07d-4ad8-a50b-c279a03a63f4)
![shifter (2)](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/137695361/a29c0dbd-4bdc-431d-9f88-5ca821c58541)

**Reply by NikderNoob95** [reply]
Thought of soldering wires to the back of the platines on the upper 2 soldering points of the switches (if looked on photo). Tested with multimeter diode the switch, if pressed, connects these points.

**Reply by NikderNoob95** [reply]
Okay looking at the top pins of the 3 they are connected inside the platine so they must be ground. Then the middle ones must be the signals.

**Reply by NikderNoob95** [reply]
And I could get the signals via the connectors. Don't have to solder anything. If you say you can help me with the code I will get me some pin-connectors that fit in the connectors on the platine to get one ground wire and 7 signal wires out of it.

**Reply by NikderNoob95** [reply]
I have to find wires like connected to the USB to TTL Converter, do you know how these are called? Since the Host shield isn't necessary anymore I could connect the wires to the Pins of the Arduino, right?

**Reply by NikderNoob95** [reply]
But I really don't need the USB PC function, and if there are wires like these with one male and one female connector, I could go directly in the Switch-Platine connectors. This would be possible to undo then if I wanted to. Luckily the connectors have the same format.

**Reply by azzajess** [reply]
Yeah you could wire directly to audrino avoiding the ones obviously used by the PS2 adaptor. I have to find out if/all pins can be connected to the audrino. 

I'll have to put together a rough script for something like this

**Reply by azzajess** [reply]
You mean something like DuPont connectors?

**Comment by NikderNoob95** [comment]
Yes thank you I found the wires and will order them. Perfect :D

**Reply by NikderNoob95** [reply]
Measured out the pins for the switches:

![Pins (1)](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/137695361/41e5c4a2-2ba2-4dc9-bc92-89f307091751)
![Pins (2)](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/137695361/d1fe7213-ed2d-4794-93c5-0bba6eb1b89f)

**Reply by azzajess** [reply]
sooo, it might explain some things, the shifter I use is a different board, with a different microcontroller thing. Not sure how its different.... Anyway, I have done investigating myself, and found out a few things. I have an unused 7th gear lol. This will be a side project but I should make up these different face plates in case a replacement is needed, or in this case, add a 7th gear... Maybe thats the difference, because your is a 6+R where as mine is a 6+R+Switch. Which makes it more similar to the 7+R versions? Anyway. My pinouts are different but fundamentally work the same. Lets see how coding this goes...
![Pinout of Shifter PCB 6+R+switch Ver](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/40465008/198c8cd2-cacc-424e-96a9-0270fc47ff4d)
![Shifter Microcontroller 2](https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/assets/40465008/ce5a16e1-93b1-413a-8fcc-f3dafd6da472)

**Reply by NikderNoob95** [reply]
Ah I see, yes my microcontroller has a USB-C port... Must be a different model. But cool man thank you for all this I hope this will be the solution... And I have a blocked gear, too :D Nice to know that it's maybe possible to upgrade.

**Reply by azzajess** [reply]
Try this and let me know how you go
Its pretty rough but should work? hopefully lol
https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/tree/main/HardwireShifterTest

**Reply by NikderNoob95** [reply]
Okay I wait for the cables and then I will try this out.

**Reply by NikderNoob95** [reply]
Yooo it works perfect!!! Nice :D I'll have to drive now, thank you thank you

**Reply by azzajess** [reply]
Let me know how it goes in game, hope it doesn't bug out when racing

**Reply by NikderNoob95** [reply]
Works very well in game until now, if some trouble appears I will report. :D

**Comment by azzajess** [comment]
@NikderNoob95 
Hey, I have heavily updated the code to make it more efficient etc. Is it possible for you to test it for me? 
https://github.com/azzajess/USB-Shifter-to-Thrustmaster-Wheelbase/blob/main/HardwireShifterTest/HardwireShifterTest.ino
You shouldn't have to comment anything else. Only thing you have to keep in consideration is the pins. The Pins have to go pair with the gears -1. In other words
pin 2 = gear 1
pin 3 = gear 2
pin 4 = gear 3
pin 5 = gear 4
pin 6 = gear 5
pin 7 = gear 6
pin 8 =  gear 7(ignore if dont have)
pin 9 = Reverse/gear 8
pin 10 = toggle switch for sequential mode (ignore if dont have)

**Reply by NikderNoob95** [reply]
Tried it out, works perfectly fine! I just had to switch the Pin for the reverse gear to 9. Thank you! With the old code the signal in serial monitor was refreshing constantly, now that stopped and it shows a clear signal. cool!

**Reply by azzajess** [reply]
Awesome thanks for testing this out

**Reply by NikderNoob95** [reply]
Hi, I've been driving for a while now and it is a complete new experience... So much fun. But I came to an idea while driving. Whenever I shift the gears, there is a latency like when the switch is activated, the driver in game starts to shift. Maybe a Gt7 specific problem but could there be a way to get this smooth? Like that it detects the shift into neutral without having to press the clutch?

**Reply by azzajess** [reply]
A couple of questions/suggestions
If you feel there is a delay, you can try lowering the `const int debounceDelay = 50;` to a smaller number. Just check to make sure that you dont get ghost triggers when switching gears.

Another thing that I think may be what you experiencing is in GT7 some cars behave weirdly with a shifter. From memory, vehicles that had a manual gearbox in the vehicle, worked perfectly fine with shifter, clutch etc. But more modern cars that used paddle shifter instead, behaved weirdly when trying to shift though gears. I ended up just using sequential shifting or paddle shifting for those cars that didn't use a shifter.

If this isnt what your talking about, give me more info on the issue your experiencing? Are you saying you want it to shift gear without having to use clutch or something?

**Reply by NikderNoob95** [reply]
Okay yes I think it is a problem specific with Gt7, I'm playing in VR so I see the drivers arm, I tried it again today and the shifting works perfect with manual cars, I drive the sequential ones with my paddles anyway. It's just in VR that the driver animation will start when shifting is done so I think theres nothing to do about it but I can live with it. :D Hopefully they fix this.
